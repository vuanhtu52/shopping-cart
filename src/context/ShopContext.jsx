import { createContext } from "react";

const ShopContext = createContext({
    products: [],
    cartItems: {},
    addToCart: () => {},
    addMultipleToCart: () => {},
    removeOneFromCart: () => {},
    removeFromCart: () => {},
    updateCart: () => {},
    getTotalCartItems: () => {},
    getTotalCartPrice: () => {},
});

export default ShopContext;