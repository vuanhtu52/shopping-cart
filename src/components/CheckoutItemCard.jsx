import { useContext } from "react";
import ShopContext from "../context/ShopContext";
import { Link } from "react-router-dom";
import BinIcon from "../assets/svg/bin.svg";

const CheckoutItemCard = ({ productId }) => {
    const { products, cartItems, addToCart, removeOneFromCart, removeFromCart, updateCart } = useContext(ShopContext);

    const getProduct = () => {
        for (const product of products) {
            if (product.id === Number(productId)) {
                return product;
            }
        }

        return {};
    };

    const product = getProduct();

    const handleClickDecrement = () => {
        removeOneFromCart(productId);
    };

    const handleClickIncrement = () => {
        addToCart(productId);
    };

    const handleInputChange = e => {
        const newValue = Number(e.target.value);

        if (newValue > 1000) {
            updateCart(productId, 1000);
        } else if (newValue < 0) {
            updateCart(productId, 0);
        } else {
            updateCart(productId, newValue);
        }

    };

    const handleClickDelete = () => {
        removeFromCart(productId);
    };

    return (
        <div className="flex w-full gap-8 border-b border-gray-300">
            <Link to={`/product/${productId}`} className="w-1/6 flex justify-center items-center">
                <img
                    src={product.image}
                    alt={product.title}
                    className="object-contain min-w-32 w-32 p-4"
                />
            </Link>
            <div className="w-2/3 flex flex-col gap-4 items-start justify-center">
                <Link to={`/product/${productId}`} className="font-bold hover:text-[#fb923c]">{product.title}</Link>
                <div className="flex gap-4 justify-center items-center">
                    {
                        cartItems[productId] > 1 ?
                            <button className="w-5 h-5 flex justify-center items-center" onClick={handleClickDecrement}>-</button> :
                            <button data-testid="decrement-button" className="w-5 h-5" onClick={handleClickDelete}>
                                <img src={BinIcon} />
                            </button>
                    }
                    <input
                        className="min-w-24 text-center border border-[#d6d3d1] [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
                        type="number"
                        min="0"
                        max="1000"
                        value={cartItems[productId]}
                        onChange={handleInputChange}
                    />
                    <button className="w-5 h-5 flex justify-center items-center" onClick={handleClickIncrement}>+</button>
                </div>
                <button data-testid="delete-button" className="text-sm text-[#ef4444] hover:underline" onClick={handleClickDelete}>Delete</button>
            </div>
            <div data-testid="total-price" className="w-1/6 flex justify-center items-center">{parseFloat((product.price * cartItems[productId]).toFixed(2))}</div>
        </div>
    );
}

export default CheckoutItemCard;