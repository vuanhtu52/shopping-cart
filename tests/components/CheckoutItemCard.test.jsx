import { MemoryRouter } from "react-router-dom";
import { describe, vi } from "vitest";
import ShopContext from "../../src/context/ShopContext";
import CheckoutItemCard from "../../src/components/CheckoutItemCard";
import { fireEvent, render, screen, within } from "@testing-library/react";

const mockProduct = {
    id: 1,
    title: "Test Product",
    price: 10,
    image: "assets/jpg/electronics.jpg"
};

const mockContext = {
    products: [mockProduct],
    cartItems: {1: 2},
    addToCart: vi.fn(),
    addMultipleToCart: vi.fn(),
    removeOneFromCart: vi.fn(),
    removeFromCart: vi.fn(),
    updateCart: vi.fn(),
    getTotalCartItems: vi.fn(),
    getTotalCartPrice: vi.fn(),
};

const renderComponent = () => {
    return render(
        <MemoryRouter>
            <ShopContext.Provider value={mockContext}>
                <CheckoutItemCard productId={1}></CheckoutItemCard>
            </ShopContext.Provider>
        </MemoryRouter>
    );
};

describe("CheckoutItemCard component", () => {
    it("Component renders correctly", () => {
        const {container} = renderComponent();
        expect(container).toMatchSnapshot();
    });

    it("Component renders product info correctly", () => {
        renderComponent();

        expect(screen.getByText(mockProduct.title)).toBeInTheDocument();
        expect(screen.getByRole("img")).toHaveAttribute("src", "assets/jpg/electronics.jpg");   // Image
        expect(screen.getByRole("spinbutton")).toHaveValue(2);  // Quantity
        expect(screen.getByText("20")).toBeInTheDocument(); // Price
    });

    it("Clicking the increment button increases quantity and price", () => {
        renderComponent();

        const incrementButton = screen.getByText("+");
        const quantityInputField = screen.getByRole("spinbutton");
        const totalPriceDisplay = screen.getByTestId("total-price");

        expect(quantityInputField).toHaveValue(2);
        expect(within(totalPriceDisplay).getByText("20")).toBeInTheDocument();

        fireEvent.click(incrementButton);

        expect(mockContext.addToCart).toHaveBeenCalledTimes(1);
        expect(mockContext.addToCart).toHaveBeenCalledWith(1);
    });

    it("Clicking the decrement button decreases quantity and price", () => {
        renderComponent();

        const decrementButton = screen.getByText("-");
        const quantityInputField = screen.getByRole("spinbutton");
        const totalPriceDisplay = screen.getByTestId("total-price");

        expect(quantityInputField).toHaveValue(2);
        expect(within(totalPriceDisplay).getByText("20")).toBeInTheDocument();

        fireEvent.click(decrementButton);

        expect(mockContext.removeOneFromCart).toHaveBeenCalledTimes(1);
        expect(mockContext.removeOneFromCart).toHaveBeenCalledWith(1);
    });

    it("Clicking the decrement button when quantity is 1 removes the item", () => {
        mockContext.cartItems[1] = 1;
        renderComponent();

        const decrementButton = screen.getByTestId("decrement-button");

        fireEvent.click(decrementButton);

        expect(mockContext.removeFromCart).toHaveBeenCalledTimes(1);
        expect(mockContext.removeFromCart).toHaveBeenCalledWith(1);
    });

    it("Entering quantity value in the input field updates the quantity and price accordingly", () => {
        renderComponent();

        const input = screen.getByRole("spinbutton");

        fireEvent.change(input, {target: {value: "500"}});

        expect(mockContext.updateCart).toHaveBeenCalledTimes(1);
        expect(mockContext.updateCart).toHaveBeenCalledWith(1, 500);
    });

    it("Entering quantity value greater than 1000 sets the quantity to 1000", () => {
        renderComponent();

        const input = screen.getByRole("spinbutton");

        fireEvent.change(input, {target: {value: "1500"}});

        expect(mockContext.updateCart).toHaveBeenCalledWith(1, 1000);
    });

    it("Entering quantity value less than 0 sets the quantity to 0", () => {
        renderComponent();

        const input = screen.getByRole("spinbutton");

        fireEvent.change(input, {target: {value: "-100"}});

        expect(mockContext.updateCart).toHaveBeenCalledWith(1, 0);
    });

    it("Clicking delete button removes the item", () => {
        renderComponent();

        const deleteButton = screen.getByTestId("delete-button");

        fireEvent.click(deleteButton);

        expect(mockContext.removeFromCart).toHaveBeenCalledWith(1);
    });
});