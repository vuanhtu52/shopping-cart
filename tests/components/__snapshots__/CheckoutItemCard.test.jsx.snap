// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckoutItemCard component > Component renders correctly 1`] = `
<div>
  <div
    class="flex w-full gap-8 border-b border-gray-300"
  >
    <a
      class="w-1/6 flex justify-center items-center"
      href="/product/1"
    >
      <img
        alt="Test Product"
        class="object-contain min-w-32 w-32 p-4"
        src="assets/jpg/electronics.jpg"
      />
    </a>
    <div
      class="w-2/3 flex flex-col gap-4 items-start justify-center"
    >
      <a
        class="font-bold hover:text-[#fb923c]"
        href="/product/1"
      >
        Test Product
      </a>
      <div
        class="flex gap-4 justify-center items-center"
      >
        <button
          class="w-5 h-5 flex justify-center items-center"
        >
          -
        </button>
        <input
          class="min-w-24 text-center border border-[#d6d3d1] [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
          max="1000"
          min="0"
          type="number"
          value="2"
        />
        <button
          class="w-5 h-5 flex justify-center items-center"
        >
          +
        </button>
      </div>
      <button
        class="text-sm text-[#ef4444] hover:underline"
        data-testid="delete-button"
      >
        Delete
      </button>
    </div>
    <div
      class="w-1/6 flex justify-center items-center"
      data-testid="total-price"
    >
      20
    </div>
  </div>
</div>
`;
